deeppavlov/core/data/data_fitting_iterator.py

```diff
@@ -0,0 +1,68 @@
+"""
+Copyright 2017 Neural Networks and Deep Learning lab, MIPT
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+from random import Random
+from typing import List, Generator, Tuple, Any
+
+from deeppavlov.core.common.registry import register
+from deeppavlov.core.common.log import get_logger
+
+logger = get_logger(__name__)
+
+
+@register('data_fitting_iterator')
+class DataFittingIterator:
+    """
+    Dataset iterator for fitting models, e. g. vocabs, kNN, vectorizers.
+    Data is passed as a list of strings(documents).
+    Generate batches (for large datasets).
+    """
+
+    def __init__(self, data: List[str], doc_ids: List[Any]=None,
+                 seed: int = None, shuffle: bool = None, batch_size: int = None,
+                 *args, **kwargs) -> None:
+
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+        self.random = Random(seed)
+        self.data = data
+        self.doc_index = doc_ids or self.get_doc_ids()
+        self.data = data
+
+    def get_doc_ids(self):
+        ids = [i for i in range(len(self.data))]
+        return ids
+
+    def gen_batch(self, batch_size=1000, shuffle=False) -> Generator[Tuple[List[list],
+                                                                           List[int]], Any, None]:
+        _batch_size = self.batch_size or batch_size
```

Comment:
Argument to the method should have priority. One could set default value here to `None` and then
```
_batch_size = batch_size or self.batch_size or 1000
```
or just have default value for self.batch_size

Reply:
Disagree.
When the instance is initialized, a user sets all needed arguments values (which are always None by default), and then these values are used in all methods. If they do not set the initial values, the values from the methods are taken. I could set default __init__ values to the methods' values, but I'd like to give a user an ability to use these methods standalone (though it is a very rare use case).
If anybody sets the `batch_size` or `shuffle` to None, then an error raises, and this is a proper behavior.

Reply:
When I specifically ask for `iterator.gen_batch(batch_size=3)` and get batches of 10, I'd think that something is wrong.

Reply:
If you want `batch_size = 3`, you should init the class with `batsh_size=3`.  And if you have a use case when you need to use `gen_batch()` always with different `batch_size`, then leave in init the default value, which is None.

Reply:
I agree with @yoptar When `get batch(batch_size=3)` return batches of size other then 3 it's not appropriate behavior. Moreover, DataFitingIterator and DataLearningIterator should have consistent behavior: get batch(batch_size=3) should work exactly the same way on both iterators.

Reply:
```
# proper design:

class MyIterator:
    def __init__(self, data, batch_size=None):
        self.data = data
        self._batch_size = batch_size

    @property
    def batch_size(self):
        return self._batch_size

    @batch_size.setter
    def batch_size(self, batch_size):
        self._batch_size = batch_size

    def gen_batches(self):
        batches = [] # count with self._batch_size
        for i, doc_ids in enumerate(batches):
            pass # whatever


iterator = MyIterator(["data"], 23)
iterator.batch_size = 12
docs = iterator.gen_batches()


# inconvenient design:

class MyIterator:
    def __init__(self, data):
        self.data = data

    @property
    def batch_size(self):
        return self._batch_size

    @batch_size.setter
    def batch_size(self, batch_size):
        self._batch_size = batch_size

    def gen_batches(self, batch_size):
        batches = []  # count with batch_size
        for i, doc_ids in enumerate(batches):
            pass  # whatever


iterator = MyIterator(["data"])
docs = iterator.gen_batches(12)
```

Reply:
`DataLearningIterator` is designed inconveniently, like the second one in the snippet above.  Instead of initializing the object once with proper parameters, and reassign them safely when needed, I have to mess up with bool and None checks inside the methods, and keep in mind what value of `batch_size` I have now and assign the same `batch_size` value in many places in my code to make sure that it is the same.

Reply:
Agree. One way to eliminate inconsistent behavior is to remove `batch_size` parameter from `gen_batch` and use it only in `__init__` for both iterators: `data_learning_iterator` and `data_fitting_iterator`. I will accept such refactoring.

Reply:
Ok then, I'll refactor both classes, run tests and update the pull-request.

Reply:
I refactored only `DataFittingIterator` and didn't touch `DataLearningIterator` because it would affect a lot of functionality and people would become upset.
To refactor `DataLearningIterator` functionality we need not only to add properties as I did in `DataFittingIterator` but also:
1. Rewrite all inherited iterators to place `batch_size`, `shuffle`, etc. in their __init__ methods.
2. Rewrite configs to put `batch_size` not under `train_config`, but under `iterator` options.
3. Modify the `train.py` script, so there would be proper properties assigning and calling.